Add documentation for Kotlin Property assignment (#24021)
gradle/gradle: subprojects/docs/src/docs/release/notes.md

More details can be found in the [Gradle Wrapper](userguide/gradle_wrapper.html#sec:adding_wrapper) section.

### Kotlin DSL

#### Experimental lazy property assignment for Kotlin scripts

> It is now possible to use the `=` operator to assign values to `Property` types in Kotlin scripts as an alternative to the `set()` method:

Reviewer: 💭 
Is there a benefit for one way or the other? Why would we want to change?

Reviewer 2: This is detailed just below the example.

```kotlin
interface Extension {
    val description: Property<String>
}

extension {
    // Using the `set()` method call
    description.set("Hello Property")
    // Gradle 8.1+ with lazy property assignment enabled
    description = "Hello Property"
}
```

This reduces the verbosity of Kotlin DSL when [lazy property types](userguide/lazy_configuration.html#lazy_properties) are used to configure tasks and extensions.
It also makes Kotlin DSL behavior consistent with Groovy DSL behavior, where using `=` to assign lazy properties has always been available.

> Lazy property assignment for Kotlin scripts is experimental opt-in feature.

Reviewer: 💄 
```suggestion
Lazy property assignment for Kotlin scripts is an experimental opt-in feature.
```

> Additionally, when value of the property is changed, project has to be reopened in the IntelliJ.

Reviewer: ```suggestion
Additionally, when the value of the property is changed, the project has to be reopened in IntelliJ.
```

More details can be found in the [Gradle Wrapper](userguide/gradle_wrapper.html#sec:adding_wrapper) section.

### Kotlin DSL

> #### Experimental Property assignment for Kotlin scripts

Reviewer 2: :x: I think we should either say `Property` (as the type, with a monotype font) or just "property" (lowercase, normal font).

Reviewer 2: Sounds good.

> It's now possible to use experimental Property assignment for Kotlin scripts.

Reviewer 2: How about:

```suggestion
It is now possible to use the `=` operator to assign values to `Property` types in Kotlin scripts as an alternative to the `set()` method:
```

> This reduces the verbosity of Kotlin DSL when recommended [lazy property types](userguide/lazy_configuration.html#lazy_properties) are used in tasks and extensions.

Reviewer 2: 🤔 I would move this to after the sample code.

Also, how about:

```suggestion
This reduces the verbosity of Kotlin DSL when [lazy property types](userguide/lazy_configuration.html#lazy_properties) are used to configure tasks and extensions.
It also makes Kotlin DSL behavior consistent with Groovy DSL behavior, where using `=` to assign lazy properties has always been available.
```

```kotlin
interface Extension {
    val description: Property<String>
}

extension {

>     // Old way, with `.set()` method call (still supported)

Reviewer 2: 🤔 I wouldn't call this the "old" method.

```suggestion
    // Using the `set()` method call
```

    description.set("Hello Property")
    // Gradle 8.1+ with Property assignment enabled
    description = "Hello Property"
}
```

You can enable Kotlin assignment by adding `systemProp.org.gradle.unsafe.kotlin.assignment=true` to the `gradle.properties` file.

> For IntelliJ IDEA support that property has to be set in the project root `gradle.properties` file (also for all included builds).

Reviewer 2: ❓ Do we need to go to this much detail here, or shall we link the userguide instead?

> Additionally, when the property's value is changed, the project has to be reopened in IntelliJ.

Reviewer 2: :x: I think we shouldn't mention this in the release notes, but in the user guide.

> Property assignment is supported from IntelliJ 2022.3 and from Android Studio Giraffe.

Reviewer 2: 🤔 Do we need to mention there are bugs?

Reviewer 2: Isn't that a bit harsh? You know better than anyone how useful the IDE support is at the moment, so it's up to you, but we shouldn't overpromise (nor underpromise) on the expected experience.

Author: Removed that sentence. I think it should be good now.

> Enabling this feature will fail your build if `allWarningsAsErrors` feature is enabled for Kotlin compiler for build logic.

Reviewer 2: :x: It is enough to mention this in the user guide.
